classic/create.go

```diff
@@ -0,0 +1,254 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2014-2015 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package classic
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/ubuntu-core/snappy/arch"
+	"github.com/ubuntu-core/snappy/dirs"
+
+	"github.com/ubuntu-core/snappy/helpers"
+	"github.com/ubuntu-core/snappy/progress"
+	"github.com/ubuntu-core/snappy/release"
+)
+
+var (
+	lxdBaseURL   = "https://images.linuxcontainers.org"
+	lxdIndexPath = "/meta/1.0/index-system"
+)
+
+func findDownloadPathFromLxdIndex(r io.Reader) (string, error) {
+	arch := arch.UbuntuArchitecture()
+	lsb, err := release.ReadLsb()
+	if err != nil {
+		return "", err
+	}
+	release := lsb.Codename
+
+	needle := fmt.Sprintf("ubuntu;%s;%s;default", release, arch)
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		if strings.HasPrefix(scanner.Text(), needle) {
+			l := strings.Split(scanner.Text(), ";")
+			if len(l) < 6 {
+				return "", fmt.Errorf("can not find download path in %s", scanner.Text())
+			}
+			return l[5], nil
+		}
+	}
+
+	if err := scanner.Err(); err != nil {
+		return "", err
+	}
+
+	return "", fmt.Errorf("needle %s not found", needle)
+}
+
+func findDownloadURL() (string, error) {
+	resp, err := http.Get(lxdBaseURL + lxdIndexPath)
+	if err != nil {
+		return "", fmt.Errorf("failed to downlaod lxdIndexUrl: %s", err)
+	}
+	defer resp.Body.Close()
+
+	dlPath, err := findDownloadPathFromLxdIndex(resp.Body)
+	if err != nil {
+		return "", err
+	}
+	url := fmt.Sprintf("%s%s%s", lxdBaseURL, dlPath, "rootfs.tar.xz")
+
+	return url, nil
+}
+
+func downloadFile(url string, pbar progress.Meter) (string, error) {
+	name := "classic"
+
+	w, err := ioutil.TempFile("", name)
+	if err != nil {
+		return "", err
+	}
+	defer func() {
+		if err != nil {
+			os.Remove(w.Name())
+		}
+	}()
+	defer w.Close()
+
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return "", err
+	}
+
+	client := &http.Client{}
```

Comment:
Thanks! I was not aware of this, added a 30s default timeout but happy to tweak the value. Firefox uses a value of 300s here but that seems a bit excessive especially since the downloader in classic will have to download around ~80MB of data.